sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -250,6 +250,7 @@ private static final class SharedStateHolder {
 
   private transient @Nullable LoadingCache<KafkaSourceDescriptor, AverageRecordSize>
       avgRecordSizeCache;
+  private @Nullable KafkaMetrics kafkaResults;
```

Comment:
instead of using a member variable how about passing this as a parameter to poll?  Then it can just be on the stack in processElement and doesn't need to be marked Nullable so it would remove some checks.